knimin/lib/data_access.py

```diff
@@ -2716,17 +2705,288 @@ def delete_samples(self, ids):
+                    raise ValueError('Sample ID %s cannot be deleted because '
+                                     'it is associated with sample plate(s) %s'
+                                     '.' % (sample_id, sample_plate_ids))
+                sql = """DELETE FROM pm.sample
+                         WHERE sample_id = %s
                          RETURNING sample_id"""
                 TRN.add(sql, [sample_id])
 
+    def _sample_plate_exists(self, id):
+        """Confirms that a sample plate ID exists
+
+        Parameters
+        ----------
+        id : int
+
+        Raises
+        ------
+        ValueError
+            If the sample plate ID does not exist
+        """
+        sql = """SELECT EXISTS (SELECT 1 FROM pm.sample_plate
+                                WHERE sample_plate_id = %s)"""
+        if not self._con.execute_fetchone(sql, [id])[0]:
+            raise ValueError('Sample plate ID %s does not exist.' % id)
+
+    def _sample_plate_is_unique(self, name, skip_id=None):
+        """Confirms that a sample plate name is not duplicate
+
+        Parameters
+        ----------
+        name : str
+        skip_id : int, optional
+            Skip this ID in searching
+
+        Raises
+        ------
+        ValueError
+            If the sample plate name already exists
+        """
+        sql = """SELECT sample_plate_id
+                 FROM pm.sample_plate
+                 WHERE name = %s
+                 AND sample_plate_id IS DISTINCT FROM %s"""
+        res = self._con.execute_fetchone(sql, [name, skip_id])
+        if res:
+            raise ValueError('Name %s conflicts with exisiting sample plate '
+                             '%s.' % (repr(name), res[0]))
+
+    def _email_exists(self, email):
+        """Confirms that an email exists
+
+        Parameters
+        ----------
+        email : str
+
+        Raises
+        ------
+        ValueError
+            If the email does not exist
+        """
+        sql = """SELECT EXISTS (SELECT 1 FROM ag.labadmin_users
+                                WHERE email = %s)"""
+        if not self._con.execute_fetchone(sql, [email])[0]:
+            raise ValueError('Email %s does not exist.' % email)
+
+    def _plate_type_name_to_id(self, name=None):
+        """Gets plate type ID by name
+
+        Parameters
+        ----------
+        name : str, optional
+            Plate type name. If None, the first plate type will be selected
+
+        Returns
+        -------
+        int
+            ID of the plate type
+
+        Raises
+        ------
+        ValueError
+            If the plate type name does not exist
+        """
+        if name:
+            sql = """SELECT plate_type_id
+                     FROM pm.plate_type
+                     WHERE name = %s"""
+            res = self._con.execute_fetchone(sql, [name])
+            if not res:
+                raise ValueError('Plate type %s does not exist.' % repr(name))
+            return res[0]
```

Comment:
this `return` only needs to be done once and can be done outside of the `if` block